qiskit/circuit/quantumcircuit.py

```diff
@@ -2216,19 +2217,35 @@ def remove_final_measurements(self, inplace: bool = True) -> Optional["QuantumCi
         dag = circuit_to_dag(circ)
         remove_final_meas = RemoveFinalMeasurements()
         new_dag = remove_final_meas.run(dag)
-
-        # Set circ cregs and instructions to match the new DAGCircuit's
+        kept_cregs = set(new_dag.cregs.values())
+        kept_clbits = set(new_dag.clbits)
+
+        # Filter only cregs/clbits still in new DAG, preserving original circuit order
+        cregs_to_add = [creg for creg in circ.cregs if creg in kept_cregs]
+        clbits_to_add = [clbit for clbit in circ._clbits if clbit in kept_clbits]
```

Comment:
Do you mean specifically relying on `circuit_to_dag` preserving the ordering of registers and bits? I actually did this because I hadn't realized `DAGCircuit.cregs` was an `OrderedDict`, but I haven't checked if `circuit_to_dag` respects order.